File: scripts/dc_probe_gz_user.mjs
/**
 * Probe Great Zeong damage with user-reported stats.
 * Run: node scripts/dc_probe_gz_user.mjs  (server on :5050)
 */
const BASE = 'http://127.0.0.1:5050';

async function fetchJson(path) {
  const r = await fetch(`${BASE}${path}`);
  return r.json();
}

const F = Math.floor;
const C = Math.ceil;
const MX = Math.max;
const EXP = Math.exp;
const DC_SHEET_UNIT_STAT_RATIO = true;

function calcStepByStep({
  unitAtk,
  charAtk,
  charDef,
  unitDef,
  wp,
  N,
  defDebuffPct = 0,
}) {
  const terrainCorrection = 1;
  const characterStatRatio = MX(0, charAtk - charDef) / 5000;
  const unitStatRatio = MX(0, C(unitAtk / 10 - unitDef / 10)) / 5000;
  const charSigmoid = 1 / (EXP((250 * (charDef - charAtk)) / 100000) + 1);
  const unitSigmoid = 1 / (EXP((25 * (unitDef - unitAtk)) / 100000) + 1);
  const baseDamage = C((characterStatRatio + unitStatRatio + charSigmoid + unitSigmoid) * wp);
  const atkCombined = C((unitAtk + 2 * charAtk) / 10);
  const defCombined = C((unitDef + 2 * charDef) / 10);
  const offExp = ((5000 - atkCombined) * 30) / 100000;
  const defExp = ((5000 - defCombined) * 3) / 100000;
  const offenseComponent = 100 / (EXP(offExp) + 1);
  const defenseComponent = -40 / (EXP(defExp) + 1);
  const offenseCorrection = C(offenseComponent * baseDamage);
  const defenseCorrection = C(defenseComponent * baseDamage);
  const damageCorrection = (offenseComponent + defenseComponent) * baseDamage;
  let battleDamage = C((baseDamage + damageCorrection) * terrainCorrection);
  const battleDamageSplit = C((baseDamage + offenseCorrection + defenseCorrection) * terrainCorrection);
  const scaledNormal = C((N * battleDamage) / 100);
  const scaledNormalSplit = C((N * battleDamageSplit) / 100);
  const normalDmg = C(battleDamage + scaledNormal);
  const normalDmgSplit = C(battleDamageSplit + scaledNormalSplit);
  return {
    baseDamage,
    battleDamage,
    battleDamageSplit,
    scaledNormal,
    scaledNormalSplit,
    normalDmg,
    normalDmgSplit,
    atkCombined,
    defCombined,
    offenseCorrection,
    defenseCorrection,
    damageCorrection: C(damageCorrection),
  };
}

async function main() {
  const gz = await fetchJson('/api/unit/1850001650?lang=EN&lb_tier=2');
  const pilot = await fetchJson('/api/character/1850001501?lang=EN');
  const stage = await fetchJson('/api/stage/90520001?lang=EN');
  const npc = stage.npc_details.find((n) => n.npc_id === '905200000102000003');
  const exWpn = gz.weapons.find((w) => w.is_ex);
  const exPow = exWpn?.levels?.[exWpn.levels.length - 1]?.power;
  console.log('EX weapon power from API LB2 max lv:', exPow);

  // User-reported panel values
  const user = {
    unitAtk: 16460,
    charRanged: 633,
    charAwaken: 807,
    wp: 6360,
    charDef: 705,
    unitDef: 25072,
  };

  for (const charAtk of [807, 633, Math.max(633, 807)]) {
    for (const N of [0, 12, 20, 33, 45, 53, 65]) {
      const r = calcStepByStep({ ...user, charAtk, wp: user.wp, N });
      if (r.normalDmg >= 28498 && r.normalDmg <= 28503) {
        console.log(`charAtk=${charAtk} N=${N}% → normal=${r.normalDmg} (split=${r.normalDmgSplit}) BD=${r.battleDamage}/${r.battleDamageSplit}`);
      }
    }
  }

  // Brute: find N that gives 28501 with user stats
  console.log('\n--- brute N for normalDmg=28501 ---');
  for (let N = 0; N <= 100; N++) {
    const r = calcStepByStep({ ...user, charAtk: 807, wp: user.wp, N });
    if (r.normalDmg === 28501 || r.normalDmgSplit === 28501) {
      console.log(`N=${N} normal=${r.normalDmg} split=${r.normalDmgSplit} BD=${r.battleDamage} bds=${r.battleDamageSplit}`);
    }
  }

  // Live browser calc if playwright available - skip, use evaluate via fetch to a test endpoint? 
  // Instead dump what unit atk LB2 base is
  const lb2 = gz.lb_data?.[2]?.stats_no_cond || gz.stats;
  const atkBase = lb2?.find?.((s) => s.name === 'Attack')?.total;
  console.log('\nAPI LB2 base Attack:', atkBase);
  console.log('User unitAtk 16460 implies +6681 bonus → base ~', 16460 - 6681);
}

main().catch((e) => {
  console.error(e);
  process.exit(1);
});

File: scripts/dc_probe_gz_user.py
"""Probe Great Zeong with user-reported loadout."""
import asyncio
import json
from playwright.async_api import async_playwright

BASE = "http://127.0.0.1:5050"


async def main():
    async with async_playwright() as p:
        browser = await p.chromium.launch(headless=True)
        page = await browser.new_page()
        await page.goto(f"{BASE}/", wait_until="networkidle", timeout=120_000)
        await page.wait_for_function("() => typeof calculateDamage === 'function'", timeout=120_000)

        result = await page.evaluate(
            """async () => {
              const gz = await fetch('/api/unit/1850001650?lang=EN&lb_tier=2').then(r=>r.json());
              const pilot = await fetch('/api/character/1850001501?lang=EN').then(r=>r.json());
              const stage = await fetch('/api/stage/90520001?lang=EN').then(r=>r.json());
              const npc = stage.npc_details.find(n=>n.npc_id==='905200000102000003');
              const supRes = await fetch('/api/supporters?lang=EN&page=1&per_page=100&q=33').then(r=>r.json());
              let supId = null;
              for (const row of supRes.items||[]) {
                const s = await fetch(`/api/supporter/${row.id}?lang=EN&level=100&lb_tier=3`).then(r=>r.json());
                if ((s.leader_skills||[]).some(x=>(x.desc||'').includes('33%'))) { supId = row.id; break; }
              }
              const supData = supId ? await fetch(`/api/supporter/${supId}?lang=EN&level=100&lb_tier=3`).then(r=>r.json()) : null;

              async function runCase(label, cfg) {
                S.dc.atkUnitData = gz;
                S.dc.atkCharData = pilot;
                S.dc.defNpc = npc;
                S.dc.lbTier = 2;
                S.dc.wpnIdx = gz.weapons.findIndex(w=>w.is_ex);
                const ex = gz.weapons[S.dc.wpnIdx];
                S.dc.wpnLv = ex.levels.length - 1;
                S.dc.optionParts = [{ details: 'Increase Attack by 12%' }];
                S.dc.supporters = supData ? [{ ...supData, _dcLevel:100, _dcLbTier:3 }] : [];
                S.dc.dtuPhysical = cfg.dtu ?? 35;
                S.dc.dmgIncrease = cfg.dmgInc ?? 0;
                S.dc.critDmgUp = 0;
                S.dc.mpLevel = cfg.mp ?? 'medium';
                S.dc.terrain = 0;
                S.dc.defending = false;
                S.dc.defNpcMapBonusesOn = false;
                S.dc.unitCondPassive = !!cfg.ucp;
                S.dc.charCondPassive = !!cfg.ccp;
                S.dc.supportCounterAtk = !!cfg.sac;
                S.dc.finalWpnPow = 0;
                S.dc.exSquadAtkPct = 0;
                S.dc.squadCondPct = 0;
                if (S.dc.atkCharData && !S.dc.atkCharData._manual) _dcRecalcPilotBonuses(true);
                const r = calculateDamage();
                return { label, cfg, normalDmg: r.normalDmg, battleDamage: r.battleDamage, baseDamage: r.baseDamage,
                  unitAtk: r.unitAtk, charAtk: r.charAtk, charDef: r.charDef, weaponPower: r.weaponPower,
                  totalNormalMultPct: r.totalNormalMultPct, scaledNormal: r.scaledNormal,
                  userDmgIncreasePct: r.userDmgIncreasePct, supportCounterAtkPctApplied: r.supportCounterAtkPctApplied };
              }

              const cases = [
                ['baseline med vigor no CP', { mp:'medium', ccp:0, ucp:0, sac:0 }],
                ['super vigor no CP', { mp:'super', ccp:0, ucp:0, sac:0 }],
                ['super vigor CP on', { mp:'super', ccp:1, ucp:0, sac:0 }],
                ['med vigor CP on (wrong?)', { mp:'medium', ccp:1, ucp:0, sac:0 }],
                ['super CP + support counter', { mp:'super', ccp:1, ucp:0, sac:1 }],
                ['med no CP + support counter', { mp:'medium', ccp:0, ucp:0, sac:1 }],
                ['med no CP + sac + dmgInc 20', { mp:'medium', ccp:0, ucp:0, sac:1, dmgInc:20 }],
              ];
              const out = [];
              for (const [label, cfg] of cases) out.push(await runCase(label, cfg));
              return out;
            }"""
        )
        print(json.dumps(result, indent=2))
        await browser.close()


if __name__ == "__main__":
    asyncio.run(main())
